src/topotoolbox/grid_object.py

```diff
@@ -516,6 +516,52 @@ def curvature(self, ctype='profc', meanfilt=False) -> 'GridObject':
         result.z = curvature
         return result
 
+    def erode(
+            self, size: tuple | None = None, footprint: np.ndarray | None = None,
+            structure: np.ndarray | None = None) -> 'GridObject':
+        """Apply a morphological erosion operation to the GridObject. Either
+        size, footprint or structure has to be passed to this function.
+        Otherwise, a default size will be used.
+
+        Parameters
+        ----------
+        size : tuple of ints
+            A tuple of ints containing the shape of the structuring element.
+            Only needed if neither footprint nor structure is provided.
+            Defaults to (3,3)
+        footprint : np.ndarray of ints, optional
+            A boolean array defining the footprint of the erosion operation.
+            Non-zero elements define the neighborhood over which the erosion
+            is applied. If None, the default is a full 
+            connectivity neighborhood.
+        structure : np.ndarray of ints, optional
+            A structuring element used for the erosion. This defines the 
+            connectivity of the elements. If None, a flat structuring element
+            is used.
+
+        Returns
+        -------
+        GridObject
+            A GridObject storing the computed values."""
+
+        # Replace NaN values with inf
+        dem = self.z.copy()
+        dem[np.isnan(dem)] = np.inf
+
+        if size is None and (structure is None and footprint is None):
+            # TODO: choose default vaule
+            size = (3, 3)
```

Comment:
I think rather than setting a default size, we should throw an error if you don't provide any structuring element. MATLAB TopoToolbox, MATLAB's `imerode` function and scipy's `grey_erosion` all throw an error in this case. There are so many possibilities here -- different sizes, non-flat structuring elements, flat but non-zero structuring elements, structuring elements that aren't square -- that users really should specify what they mean.

This error message should be informative. Something like "erode requires a structuring element to be specified. Use the size argument for a full and flat structuring element (equivalent to a a minimum filter) or the structure and footprint arguments to specify a non-flat structuring element."

Reply:
Done :+1: